src/evolve.ts

```diff
@@ -0,0 +1,236 @@
+type GetValueByKey<
+  T,
+  K extends PropertyKey,
+  Default = never,
+> = K extends keyof T ? T[K] : Default;
+
+/**
+ * Get the type of first parameter.
+ * But get `never` if the function take two or more required parameters.
+ * @example
+ * type Number = GetFirstParam<(arg1: number) => void>; // type Number = number
+ * type Number = GetFirstParam<(arg1: number, arg2?: number) => void>; // type Number = number
+ * type Number = GetFirstParam< (arg1: number, arg2?: number, arg3?: number) => void >; // type Number = number
+ * type Never = GetFirstParam<(arg1: number, arg2: number | undefined) => void>; // type Never = never
+ */
+type GetFirstParam<T extends AFunction> = T extends (
+  firstArg: infer Ret,
+  ...restArgs: infer Rest
+) => any
+  ? 0 extends Rest['length']
+    ? Ret
+    : never
+  : never;
+
+/**
+ * Creates an assumed `data` type from the type of the `evolver` argument.
+ * @example
+ * type Number = EvolveTarget<(arg1: number) => void>; // type Number = number
+ * type Nested = EvolveTarget<{
+ *   num: (arg1: number) => void;
+ *   obj: {
+ *    num: (arg1: number) => void
+ *   }
+ *   notFunc: undefined
+ * }>;
+ * // type Nested = {
+ * //     num?: number | undefined;
+ * //     obj?: {
+ * //         num?: number | undefined;
+ * //     } | undefined;
+ * // }
+ */
+type EvolveTarget<E> = E extends AFunction
+  ? GetFirstParam<E>
+  : E extends object
+    ? {
+        [K in keyof E as E[K] extends AFunction
+          ? K
+          : E[K] extends object
+            ? K
+            : never]?: EvolveTarget<E[K]>;
+      }
+    : never;
+
+/**
+ * Creates an assumed `evolver` type from the type of `data` argument.
+ * Note: `[T] extends [Something] ? ...` is conditional type avoiding distributivity.
+ * @example
+ * interface Data {
+ *   id: number;
+ *   quartile: Array<number>;
+ *   time?: { elapsed: number; remaining?: number };
+ * }
+ * type Nested = Evolver<Data>; // type Nested = {
+ * //  id?: ((data: number) => any) | undefined;
+ * //  quartile?: ((data: number[]) => any) | undefined;
+ * //  time?:
+ * //    | ((
+ * //        data:
+ * //          | {
+ * //              elapsed: number;
+ * //              remaining?: number | undefined;
+ * //            }
+ * //          | undefined
+ * //      ) => any)
+ * //    | {
+ * //        elapsed?: ((data: number) => any) | undefined;
+ * //        remaining?: ((data: number | undefined) => any) | undefined;
+ * //      }
+ * //    | undefined;
+ * //};
+ */
+type Evolver<T> = [T] extends [Array<any>]
+  ? (data: T) => any
+  : T extends object
+    ? {
+        [K in keyof T]?:
+          | ([T[K]] extends [Array<any>] ? never : (data: T[K]) => any)
+          | Evolver<T[K]>;
+      }
+    : never;
+
+/**
+ * Creates return type from the type of arguments of `evolve`.
+ */
+type Evolved<T, E> = E extends AFunction
+  ? ReturnType<E>
+  : T extends object
+    ? {
+        [K in keyof T]: Evolved<T[K], GetValueByKey<E, K, T[K]>>;
+      }
+    : T;
```

Comment:
And similarly this type doesn't need to be this complicated:

```suggestion
type Evolved<T, E> = T extends object
  ? {
      [K in keyof T]: K extends keyof E
        ? E[K] extends (...arg: any) => infer R
          ? R
          : Evolved<T[K], E[K]>
        : T[K];
    }
  : T;
```